src/clj/harja/palvelin/integraatiot/api/tyomaapaivakirja.clj

```diff
@@ -0,0 +1,255 @@
+(ns harja.palvelin.integraatiot.api.tyomaapaivakirja
+  "Työmaapäiväkirjan hallinta API:n kautta. Alkuun kirjaus ja päivitys mahdollisuudet"
+  (:require [com.stuartsierra.component :as component]
+            [clojure.spec.alpha :as s]
+            [compojure.core :refer [POST PUT]]
+            [harja.kyselyt.tyomaapaivakirja :as tyomaapaivakirja-kyselyt]
+            [harja.kyselyt.konversio :as konv]
+            [harja.palvelin.integraatiot.api.tyokalut.json :refer [pvm-string->java-sql-date aika-string->java-sql-date]]
+            [harja.palvelin.komponentit.http-palvelin :refer [julkaise-reitti poista-palvelut]]
+            [harja.palvelin.integraatiot.api.tyokalut.kutsukasittely :refer [kasittele-kutsu]]
+            [harja.palvelin.integraatiot.api.tyokalut.json-skeemat :as json-skeemat]
+            [harja.palvelin.integraatiot.api.validointi.parametrit :as parametrivalidointi]
+            [harja.palvelin.integraatiot.api.tyokalut.virheet :as virheet]
+            [clojure.java.jdbc :as jdbc])
+  (:use [slingshot.slingshot :only [throw+]]))
+
+(s/def ::urakka-id #(and (integer? %) (pos? %)))
+
+(defn- tarkista-parametrit [parametrit]
+  (parametrivalidointi/tarkista-parametrit
+    parametrit
+    {:id "Urakka-id puuttuu"})
+  (when (not (s/valid? ::urakka-id (konv/konvertoi->int (:id parametrit))))
+    (virheet/heita-viallinen-apikutsu-poikkeus
+      {:koodi virheet/+puutteelliset-parametrit+
+       :viesti (format "Urakka-id muodossa: %s. Anna muodossa: 1" (:id parametrit))})))
+
+;; TODO: Validoi sisään tuleva data
+(defn validoi-tyomaapaivakirja [data]
+  )
+
+(defn tallenna-tyomaapaivakirja [db urakka-id data kayttaja tyomaapaivakirja-id]
+  (let [tyomaapaivakirja-id (konv/konvertoi->int tyomaapaivakirja-id)
+        versio (get-in data [:tunniste :versio]) ; Jokaiselle payloadilla on oma versionsa
+        tyomaapaivakirja {:urakka_id urakka-id
+                          :kayttaja (:id kayttaja)
+                          :paivamaara (pvm-string->java-sql-date (get-in data [:tunniste :paivamaara]))
+                          :ulkoinen-id (get-in data [:tunniste :id])
+                          :id tyomaapaivakirja-id}
+
+        tyomaapaivakirja-id (if tyomaapaivakirja-id
+                              ;; Päivitä vanhaa
+                              (do
+                                (tyomaapaivakirja-kyselyt/paivita-tyomaapaivakirja<! db tyomaapaivakirja)
+                                (:id tyomaapaivakirja))
+                              ;; ELSE: Lisää uusi
+                              (:id (tyomaapaivakirja-kyselyt/lisaa-tyomaapaivakirja<! db tyomaapaivakirja)))
+
+        ;; Tallennetaan kalusto
+        kalustot (get-in data [:kaluston-kaytto])
+        _ (doall (for [k kalustot
```

Comment:
Tässä voi käyttää doseq-kutsua kaikilla for-loopin ominaisuuksilla.

Reply:
Korvasin doall - for loopit doseqillä.